pennylane/beta/tapes/qnode.py

```diff
@@ -0,0 +1,541 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the QNode class and qnode decorator.
+"""
+from collections.abc import Sequence
+from functools import lru_cache, update_wrapper
+import warnings
+
+import numpy as np
+
+from pennylane import Device
+from pennylane.beta.queuing import MeasurementProcess
+from pennylane.beta.tapes import QuantumTape
+from pennylane.beta.interfaces.autograd import AutogradInterface
+
+
+class QuantumFunctionError(Exception):
+    """Exception raised when an illegal operation is defined in a quantum function."""
+
+
+class QNode:
+    """Base class for quantum nodes in the hybrid computational graph.

```

Comment:
I have to admit, I was thinking this as well. I would be very happy to change it to `qfunc`, which I think is easier to understand.

But it's hard to avoid the history of the term 'QNode'!

(@co9olguy, thoughts?)

Reply:
It would also finally explain why a `QNode` raises a `QuantumFunctionError` 😆

Reply:
Also, I like `QuantumTape` but sometimes think `QuantumCircuit` might also fit :thinking: Although probably too late to discuss this!

Reply:
Never too late to discuss it! At least until it's been merged into core :laughing:

I suppose there were three reasons I named it tape:

* I was still thinking in terms of queuing when I proposed it; it essentially is an evolution of the `OperatorRecorder`, recording quantum operations and measurements. 

* I was inspired by TFs gradient tape, which works very similarly - it records operations, supports nesting, and has methods for computing the gradient.

* I wanted a name that would be unique to PL and stand out :laughing: like 'qnode'.

Reply:
I agree @co9olguy. And

>  Instead of "this function represents a qnode", it is "this function contains a qnode".

is a really nice way of thinking about it.